File: the-xray-tracer/typescript/lib/the-sns-flow-stack.ts
import * as cdk from '@aws-cdk/core';
import lambda = require('@aws-cdk/aws-lambda');
import sns = require('@aws-cdk/aws-sns');
import sns_sub = require('@aws-cdk/aws-sns-subscriptions');

export interface SNSFlowStackProps extends cdk.StackProps{
    readonly snsTopicARN: string;
}

export class TheSnsFlowStack extends cdk.Stack {
    snsLambda: lambda.Function;
    constructor(scope: cdk.Construct, id: string, props: SNSFlowStackProps) {
        super(scope, id, props);

        // Create an SNS Topic
        const topic = new sns.Topic(this, 'TheXRayTracerSnsTopic', {
                displayName: "The XRay Tracer CDK Pattern Topic",
        });

        // defines an AWS Lambda resource
        this.snsLambda = new lambda.Function(this, 'snsLambdaHandler', {
            runtime: lambda.Runtime.NODEJS_12_X,
            code: lambda.Code.fromAsset('lambda-fns'),
            handler: 'sns_publish.handler',
            environment: {
                TOPIC_ARN: topic.topicArn
            },
            tracing: lambda.Tracing.ACTIVE
        });
        topic.grantPublish(this.snsLambda);
        let apigwTopic = sns.Topic.fromTopicArn(this, 'SNSTopic', props.snsTopicARN);
        apigwTopic.addSubscription(new sns_sub.LambdaSubscription(this.snsLambda));

        // Have a Lambda subscribe to our topic
        let snsSubscribeLambda = new lambda.Function(this, 'snsSubscriptionLambdaHandler', {
            runtime: lambda.Runtime.NODEJS_12_X,
            code: lambda.Code.fromAsset('lambda-fns'),
            handler: 'sns_subscribe.handler',
            tracing: lambda.Tracing.ACTIVE
        });
        topic.addSubscription(new sns_sub.LambdaSubscription(snsSubscribeLambda));
    }
}

File: the-xray-tracer/python/the_xray_tracer/the_sns_flow_stack.py
from aws_cdk import (
    aws_lambda as _lambda,
    aws_sns as sns,
    aws_sns_subscriptions as subscriptions,
    core
)


class TheSnsFlowStack(core.Stack):
    def __init__(self, scope: core.Construct, id: str, sns_topic_arn: str, **kwargs) -> None:
        super().__init__(scope, id, **kwargs)

        # SNS Topic creation
        topic = sns.Topic(self, 'TheXRayTracerSnsTopic', display_name='The XRay Tracer CDK Pattern Topic')

        sns_lambda = _lambda.Function(self, "snsLambdaHandler",
                                      runtime=_lambda.Runtime.NODEJS_12_X,
                                      handler="sns_publish.handler",
                                      code=_lambda.Code.from_asset("lambda_fns"),
                                      tracing=_lambda.Tracing.ACTIVE,
                                      environment={
                                          "TOPIC_ARN": topic.topic_arn
                                      }
                                      )
        topic.grant_publish(sns_lambda)
        apigw_topic = sns.Topic.from_topic_arn(self, 'SNSTopic', sns_topic_arn)
        apigw_topic.add_subscription(subscriptions.LambdaSubscription(sns_lambda))

        sns_subscriber_lambda = _lambda.Function(self, "snsSubscriptionLambdaHandler",
                                                 runtime=_lambda.Runtime.NODEJS_12_X,
                                                 handler="sns_subscribe.handler",
                                                 code=_lambda.Code.from_asset("lambda_fns"),
                                                 tracing=_lambda.Tracing.ACTIVE
                                                 )
        topic.add_subscription(subscriptions.LambdaSubscription(sns_subscriber_lambda))
